nupic/research/frameworks/vernon/experiments/self_supervised_experiment.py

```diff
@@ -0,0 +1,452 @@
+
+        if isinstance(dataset_class, dict):
+            dataset_class = dataset_class.get(dataset_type, None)
+
+        dataset_args = dict(dataset_args.get(dataset_type, {}))
+
+        if dataset_class is None:
+            raise ValueError("Must specify 'dataset_class' in config.")
+
+        return dataset_class(**dataset_args)
+
+    def unsupervised_loss(self, output, target, reduction="mean"):
+        """
+        The error loss component of the loss function.
+        """
+        return self._loss_function_unsupervised(output, reduction=reduction)
+
+    def supervised_loss(self, output, target, reduction="mean"):
+        """
+        The error loss component of the loss function.
+        """
+        return self._loss_function_supervised(output, target, reduction=reduction)
+
+    def encoder_complexity_loss(self, model):
+        """
+        The encoder complexity component of the loss function.
+        """
+        pass
+
+    def pre_batch_supervised(self, model, batch_idx):
+        pass
+
+    def post_batch_supervised(
+        self, model, error_loss, complexity_loss, batch_idx, num_images, time_string
+    ):
+        # Update 1cycle learning rate after every batch
+        if self.step_lr_every_batch_classifier:
+            self.lr_scheduler_classifier.step()
+
+    def post_optimizer_step_supervised(self, model):
+        pass
+
+    def post_batch_wrapper_supervised(self, **kwargs):
+        self.post_optimizer_step_supervised(self.model)
+        self.post_batch_supervised(**kwargs)
+
+    def transform_data_to_device_unsupervised(self, data, target, device, non_blocking):
+        """
+        This provides an extensibility point for performing any final
+        transformations on the data or targets. This will suffice for most
+        self-supervised experiments. The first copy of data gets passed through the
+        model, and the second copy is optionally used in the loss function.
+        """
+        data = data.to(self.device, non_blocking=non_blocking)
+        return data, data
+
+    def train_epoch(self):
+        self.train_model(
+            self.encoder,
+            self.unsupervised_loader,
+            self.encoder_optimizer,
+            self.device,
+            criterion=self.error_loss,
+            complexity_loss_fn=self.encoder_complexity_loss,
+            batches_in_epoch=self.batches_in_epoch,
+            pre_batch_callback=self.pre_batch,
+            post_batch_callback=self.post_batch_wrapper,
+            transform_to_device_fn=self.transform_data_to_device_unsupervised,
+        )
+
+    def validate(self):
+        if self.reset_classifier_on_validate:
+            for layer in self.classifier.children():
+                if hasattr(layer, "reset_parameters"):
+                    layer.reset_parameters()
+
+        for _ in range(self.supervised_training_epochs_per_validation):
+            self.train_model(
+                self.model,
```

Comment:
If I'm not mistaken, `self.model` refers to the encoder, and it's being trained within the `validate` method. Shouldn't this be `self.classifier` being trained instead (esp since you're used the supervised loader to train it)?

Reply:
In setup_experiment, the self.model initially points to encoder, but then is reassigned to the EncoderClassifier together. Therefore the forward pass goes through the encoder's encode() and then the classifier's forward().

Reply:
That seems wise.